pkg/webhook/elasticquota/quota_topology.go

```diff
@@ -0,0 +1,664 @@
+	}
+
+	// fill quota with basic Information
+	err = qt.fillDefaultQuotaInfo(quota)
+	if err != nil {
+		klog.Errorf("AddQuota fillDefaultQuotaInfo failed:%v", err)
+		return err
+	}
+
+	err = qt.basicItemCheck(quota)
+	if err != nil {
+		klog.Errorf("AddQuota basicItemCheck failed :%v", err)
+		return err
+	}
+
+	quotaInfo := core.NewQuotaInfoFromQuota(quota)
+
+	err = qt.validateQuota(quotaInfo)
+	if err != nil {
+		klog.Errorf("AddQuota validateQuota failed:%v", err)
+		return err
+	}
+
+	qt.quotaInfoMap[quotaInfo.Name] = quotaInfo
+	qt.podCache[quotaInfo.Name] = make(map[string]struct{})
+	if _, ok := qt.quotaHierarchyInfo[quotaInfo.Name]; !ok {
+		qt.quotaHierarchyInfo[quotaInfo.Name] = make(map[string]struct{})
+	}
+	if quotaInfo.ParentName != extension.RootQuotaName {
+		qt.quotaHierarchyInfo[quotaInfo.ParentName][quotaInfo.Name] = struct{}{}
+	}
+	return nil
+}
+
+func (qt *quotaTopology) UpdateQuota(oldQuota, originQuota *v1alpha1.ElasticQuota) (err error) {
+	if originQuota == nil {
+		return fmt.Errorf("AddQuota param is nil")
+	}
+
+	if oldQuota != nil && reflect.DeepEqual(quotaFieldsCopy(oldQuota), quotaFieldsCopy(originQuota)) {
+		return nil
+	}
+
+	quotaName := originQuota.Name
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+	localQuotaInfo, exist := qt.quotaInfoMap[quotaName]
+	if !exist {
+		err = fmt.Errorf("quota not exist:%v", quotaName)
+		klog.Errorf("UpdateQuota failed:%v", err)
+		return err
+	}
+
+	err = qt.basicItemCheck(originQuota)
+	if err != nil {
+		klog.Errorf("UpdateQuota basicItemCheck failed %v", err)
+		return err
+	}
+
+	newQuotaInfo := core.NewQuotaInfoFromQuota(originQuota)
+	// quota 信息检查
+	err = qt.validateQuota(newQuotaInfo)
+	if err != nil {
+		klog.Errorf("UpdateQuota validateQuota failed:%v", err)
+		return err
+	}
+
+	// quota 层次关系发生改变
+	err = qt.checkQuotaHierarchyHasChange(localQuotaInfo, newQuotaInfo)
+	if err != nil {
+		klog.Errorf("UpdateQuota checkQuotaHierarchyHasChange failed:%v", err)
+		return err
+	}
+
+	qt.quotaInfoMap[quotaName] = newQuotaInfo
+	return nil
+}
+
+func (qt *quotaTopology) DeleteQuota(quotaName string) (err error) {
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	quotaInfo, exist := qt.quotaInfoMap[quotaName]
+	if !exist {
+		return fmt.Errorf("not found quota:%v", quotaName)
+	}
+
+	// check has child quota
+	if childSet, exist := qt.quotaHierarchyInfo[quotaName]; exist {
+		if len(childSet) > 0 {
+			return fmt.Errorf("delete quota failed:%v, has child quota", quotaName)
+		}
+	} else {
+		return fmt.Errorf("BUG quotaMap and quotaTree information out of sync, losed :%v", quotaName)
+	}
+
+	// 从父quota中删除
+	if quotaInfo.ParentName != extension.RootQuotaName {
+		delete(qt.quotaHierarchyInfo[quotaInfo.ParentName], quotaName)
+	}
+	// 从拓扑结构中删除
+	delete(qt.quotaHierarchyInfo, quotaName)
+	delete(qt.quotaInfoMap, quotaName)
+	delete(qt.podCache, quotaName)
+	return nil
+}
+
+func (qt *quotaTopology) ValidateDeleteQuota(quota *v1alpha1.ElasticQuota) (err error) {
```

Comment:
what's the diff between ValidateDeleteQuota and DeleteQuota?  why deleteQuota don't check podCache?

Reply:
add check podCache logic in DeleteQuota too.

Reply:
decide to delete the ValidateDeleteQuota function. And mutatingHandler only handles the create operation, because only create operation need to fillQuotaDefaultInformation, while  validatingHandler handles create, update and delete operations, but if the operation is create, just check whether decode error or not.